### PR TITLE
Autosave projects every 30 seconds if the code has changed

### PR DESCRIPTION
**Pull request by joshlory**
Autosaves projects at most every 30 seconds.

In Blockly `dashboard.getEditorSource` is relatively expensive (because it serializes blocks to XML), so we only call it if a `workspaceChange` event has been fired since the last autosave.  In Droplet `dashboard.getEditorSource` is cheap so we call it every interval.  Once we have the code, autosave is only triggered if the code is different from what was last sent to the server.

### CODE REVIEW

File: shared/js/initApp.js [hunk 1]
```diff
@@ -104,12 +107,16 @@ var appToProjectUrl = {
...
-  $('.project_updated_at').text('Saving...');  // TODO (Josh) i18n
-  var channelId = dashboard.currentApp.id;
-  dashboard.currentApp.levelSource = window.Blockly
+dashboard.getEditorSource = function() {
+  return window.Blockly
       ? Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace))
       : window.Applab && Applab.getCode();
+};
+
+dashboard.saveProject = function(callback, overrideSource) {
```
Comment (bcjordan) [inline shared/js/initApp.js]: jsdoc comment w/ description of `overrideSource` could be handy

File: shared/js/initApp.js [hunk 2]
```diff
@@ -104,12 +107,19 @@ var appToProjectUrl = {
   applab: '/p/applab'
 };
 
-dashboard.saveProject = function(callback) {
-  $('.project_updated_at').text('Saving...');  // TODO (Josh) i18n
-  var channelId = dashboard.currentApp.id;
-  dashboard.currentApp.levelSource = window.Blockly
+/**
+ * @returns {string} The serialized app source from the editor.
```
Comment (bcjordan) [inline shared/js/initApp.js]: s/app source/level source?

File: shared/js/initApp.js [hunk 3]
```diff
@@ -1,3 +1,6 @@
+// Attempt to save projects every 30 seconds
```
Comment (bcjordan) [inline shared/js/initApp.js]: Does this apply to both "projects" and "non-projects"? If so, maybe s/`project`/`userSource` or something else that covers the full gamut of cases where used.

Comment (joshlory) [inline shared/js/initApp.js]: Just projects for now.

**Comment by bcjordan** [comment]
LGTM! Might be worth tracking (1) i18n & (2) test coverage (maybe a pair of blockly/droplet UI tests which peek at the `hasProjectChanged` variable)

File: shared/js/initApp.js (hunk 3 above)
Comment (bcjordan) [inline shared/js/initApp.js]: Ah, I see now, the code below is gated above in the method.

`initApp()` is getting (maybe always has been) pretty unwield-ily long, we should keep an eye out for opportunities to extract some named methods out of  it to make it a bit easier to read next time we're in there

**Pull request merged**